File: backend/models/ipo.py
from typing import Literal

from pydantic import BaseModel, Field


IpoCategory = Literal["mainboard", "sme"]


class IPO(BaseModel):
    id: str
    name: str
    category: IpoCategory
    status: str
    price: float | None = None
    gmp: float | None = None
    gmpPercentage: float | None = None
    estimatedListing: float | None = None
    estimatedListingPercentage: float | None = None
    expectedProfit: float | None = None
    profitType: Literal["profit", "loss"] | None = None
    lotSize: float | None = None
    issueSize: float | None = None
    subscription: float | None = None
    biddingStartDate: str | None = None
    biddingEndDate: str | None = None
    sourceUrl: str | None = None
    sourceId: str
    scrapedAt: str
    updatedAt: str
    lastSeenAt: str
    active: bool = True


class IpoGroups(BaseModel):
    mainboard: list[IPO] = Field(default_factory=list)
    sme: list[IPO] = Field(default_factory=list)


class IposResponse(BaseModel):
    success: bool
    available: bool
    data: IpoGroups
    message: str | None = None


class HealthResponse(BaseModel):
    status: Literal["ok"]

File: backend/types/ipo.ts
export type IPOCategory = "mainboard" | "sme";
export type ProfitType = "profit" | "loss";

export interface IPO {
  id: string;
  name: string;
  category: IPOCategory;
  status: string;
  price: number | null;
  gmp: number | null;
  gmpPercentage: number | null;
  estimatedListing: number | null;
  estimatedListingPercentage: number | null;
  expectedProfit: number | null;
  profitType: ProfitType | null;
  lotSize: number | null;
  issueSize: number | null;
  subscription: number | null;
  biddingStartDate: string | null;
  biddingEndDate: string | null;
  sourceUrl: string | null;
  sourceId: string;
  scrapedAt: string;
  updatedAt: string;
  lastSeenAt: string;
  active: boolean;
}

export interface ConvexIPODocument extends Omit<IPO, "id"> {
  _id?: string;
  id?: string;
}

export interface IpoGroups {
  mainboard: IPO[];
  sme: IPO[];
}

export interface IposResponse {
  success: boolean;
  available: boolean;
  data: IpoGroups;
  message?: string;
}

export function normalizeIpo(document: ConvexIPODocument): IPO {
  return {
    id: document.id ?? document._id ?? document.sourceId,
    name: document.name,
    category: document.category,
    status: document.status,
    price: document.price ?? null,
    gmp: document.gmp ?? null,
    gmpPercentage: document.gmpPercentage ?? null,
    estimatedListing: document.estimatedListing ?? null,
    estimatedListingPercentage: document.estimatedListingPercentage ?? null,
    expectedProfit: document.expectedProfit ?? null,
    profitType: document.profitType ?? null,
    lotSize: document.lotSize ?? null,
    issueSize: document.issueSize ?? null,
    subscription: document.subscription ?? null,
    biddingStartDate: document.biddingStartDate ?? null,
    biddingEndDate: document.biddingEndDate ?? null,
    sourceUrl: document.sourceUrl ?? null,
    sourceId: document.sourceId,
    scrapedAt: document.scrapedAt,
    updatedAt: document.updatedAt,
    lastSeenAt: document.lastSeenAt ?? document.updatedAt,
    active: document.active ?? true,
  };
}